### ISSUE TITLE
[2.2.x] How to disable modules per environment?

### ISSUE THREAD

**Issue by jahvi** [magento/magento2#12274]
As of 2.2 it is recommended to include the `config.php` file in version control as it provides common settings that are shared across environments. The problem is that sometimes you want to install extension only for development so if this is the case then you'll get an error if the module is included in the `config.php` file but not actually installed.

### Preconditions

1. Magento 2.2.1
2. PHP 7.0

### Steps to reproduce

1. Install fresh version of Magento 2.2.1
2. Install development only extension as a dev dependency, eg: `composer require honl/magento2-templatehints --dev`
3. Run `php bin/magento setup:upgrade`, the `config.php` file gets updated with the new extension and everything works as expected.
4. Mimic a "production" environment by deleting the `vendor` folder and running `composer install --no-dev` and deleting the `Ho_Templatehints` row in the `setup_module` table.
5. Clear cache.

### Expected result

1. Website should load normally with extension disabled.

### Actual result

1. You get a `Setup version for module 'Ho_Templatehints' is not specified` error.

The error is to be expected because Magento is trying to load a module that hasn't been installed, I though I could only enable the module in the `env.php` file since this is environment specific but that didn't work.

**Comment by renttek** [comment]
I see 2 Options to implement this requirement:

1. Add a modules key to env.php

2.Add an 'modules-{mode}' key in config.php, where mode is the run mode of magento (developer, production, default, ...)

Personally I'd prefer option 2, because you only have to look at 1 file to determine which modules are enabled/disabled in which environment/mode.

Another question is: should this be done in core?

**Comment by jahvi** [comment]
I though option 1 was already supported but it didn't work in my case, I tried manually adding the `modules` key to this file but they were not being picked up by Magento.

I personally prefer that option just because it's the way settings work now so it would make it consistent, but any fix or workaround would help. As it is now I have to install dev dependencies on the live environment which is not ideal.

**Comment by orlangur** [comment]
`--dev` in Composer should not be mixed with `developer` mode in Magento.

I believe you are supposed to have separate `config.php` for each environment if you need separate sets of modules. Generally, I don't think it's a good idea, better option is to disable unneeded modules in production (or make modules disactivate automatically in production mode).

**Comment by jahvi** [comment]
> Generally, I don't think it's a good idea, better option is to disable unneeded modules in production (or make modules disactivate automatically in production mode)

@orlangur I agree, however how can we manage to do that if the `config.php` file is encouraged to keep under version control?

Are separate `config.php` files per environment supported now? I couldn't find anything about that in the docs.

**Comment by orlangur** [comment]
> how can we manage to do that if the `config.php` file is encouraged to keep under version control?

Right, so we can have only `0` or `1` for each module which means disabled/enabled. Overlooked this.

> Are separate `config.php` files per environment supported now?

I don't think so, just that you can do it a part of your custom deployment process (copy `config.dev.php` or `config.prod.php` to `config.php`).

**Comment by jahvi** [comment]
> I don't think so, just that you can do it a part of your custom deployment process (copy config.dev.php or config.prod.php to config.php).

I see, that's a good workaround thanks for the recommendation.

**Comment by ishakhsuvarov** [comment]
@jahvi Looks like issue is solved, can we close the ticket now?

**Comment by jahvi** [comment]
Well there is a workaround, still would be nice if there was a built in way of dealing with this but this can be closed now.
